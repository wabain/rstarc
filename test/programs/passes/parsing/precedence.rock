(
    Precedence rules. The current ordering, per the spec, is:

    1. Function Call [greedy arguments]
    2. Logical NOT [right-associative]
    3. Multiplication and Division [left-associative]
    4. Addition and Subtraction [left-associative]
    5. Comparison operators [left-associative]
    6. and, or, and nor [left-associative]
)

Put 0 into U
Put 0 into V
Put 0 into W
Put 0 into X
Put 0 into Y
Put 0 into Z

Put X taking Y and Z times U plus V into Dummy
Put X plus Y minus Z into Dummy
Put X times Y over Z into Dummy

If X times Y is Z plus U
    Say nothing

If U is V times X taking Y and Z is V and W is X
    Say nothing
