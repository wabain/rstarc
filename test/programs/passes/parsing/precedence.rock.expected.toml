[tests]
enabled = ["ast"]

[ast]
returncode = "0"
stdout = """
Assign(
  U,
  0,
),
Assign(
  V,
  0,
),
Assign(
  W,
  0,
),
Assign(
  X,
  0,
),
Assign(
  Y,
  0,
),
Assign(
  Z,
  0,
),
Assign(
  Dummy,
  Add(
    Mul(
      Not(
        FuncCall(
          X,
          [
            Y,
            Z,
          ],
        ),
      ),
      U,
    ),
    V,
  ),
),
Assign(
  Dummy,
  Add(
    X,
    Sub(
      Y,
      Z,
    ),
  ),
),
Assign(
  Dummy,
  Mul(
    X,
    Div(
      Y,
      Z,
    ),
  ),
),
If(
  Comparison(
    Mul(
      X,
      Y,
    ),
    Is,
    Add(
      Z,
      U,
    ),
  ),
  [
    Say(
      null,
    ),
  ],
  [
  ],
),
If(
  And(
    Comparison(
      U,
      Is,
      Comparison(
        Mul(
          V,
          FuncCall(
            X,
            [
              Y,
              Z,
            ],
          ),
        ),
        Is,
        V,
      ),
    ),
    Comparison(
      W,
      Is,
      X,
    ),
  ),
  [
    Say(
      null,
    ),
  ],
  [
  ],
),
"""
